src/analysis/inner_product_space/of_norm.lean

```diff
@@ -0,0 +1,356 @@
+/-
+Copyright (c) 2020 Heather Macbeth. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Heather Macbeth
+-/
+
+import topology.algebra.algebra
+import analysis.inner_product_space.basic
+
+/-!
+# Inner product space derived from a norm
+
+This file defines an `inner_product_space` instance from a norm that respects the
+parallellogram identity. The parallelogram identity is a way to express the inner product of `x` and
+`y` in terms of the norms of `x`, `y`, `x + y`, `x - y`.
+
+## Main results
+
+- `inner_product_space.of_norm`: a normed space whose norm respects the parallellogram identity,
+  can be seen as an inner product space.
+
+## Tags
+
+inner product space, Hilbert space, norm
+
+## References
+
+- [Jordan, P. and von Neumann, J., *On inner products in linear, metric spaces*][Jordan1935]
+- https://math.stackexchange.com/questions/21792/norms-induced-by-inner-products-and-the-parallelogram-law
+- https://math.dartmouth.edu/archive/m113w10/public_html/jordan-vneumann-thm.pdf
+-/
+
+open is_R_or_C
+open_locale complex_conjugate
+
+/-- Predicate for the parallelogram identity to hold. -/
+class inner_product_spaceable (E : Type*) [normed_add_comm_group E] : Prop :=
+(parallelogram_identity :
+  ∀ x y : E, ‖x + y‖ * ‖x + y‖ + ‖x - y‖ * ‖x - y‖ = 2 * (‖x‖ * ‖x‖ + ‖y‖ * ‖y‖))
+
+variables (𝕜 : Type*) [is_R_or_C 𝕜] {E : Type*} [normed_add_comm_group E] [normed_space 𝕜 E]
+
+variables (𝕜)
+
+local notation `𝓚` := algebra_map ℝ 𝕜
+
+/-- Auxiliary definition of the inner product derived from the norm. -/
+private noncomputable def inner_ (x y : E) : 𝕜 :=
+4⁻¹ * ((𝓚 ‖x + y‖) * (𝓚 ‖x + y‖) - (𝓚 ‖x - y‖) * (𝓚 ‖x - y‖)
+          + (I:𝕜) * (𝓚 ‖(I:𝕜) • x + y‖) * (𝓚 ‖(I:𝕜) • x + y‖)
+          - (I:𝕜) * (𝓚 ‖(I:𝕜) • x - y‖) * (𝓚 ‖(I:𝕜) • x - y‖))
```

Comment:
Does this generalize as an approach to producing a conjugate-linear map over any star ring? My guess would be that summing over the generators does the trick...

Reply:
I don't know much about star rings. Can you detail? How are the generators denoted in mathlib?

Reply:
I guess what I probably meant here is "over the quaternions".

Reply:
Probably you're right, but I have no idea how to do it in Lean. You will need to give me more info.

Reply:
I should be clear that this comment is not intended to block this PR